### PR TITLE
Dag files

### PR DESCRIPTION
**Pull request by cimendes**
This PR addresses an issue raised in #194 where the .treeDag.json and forktree.json files aren't automatically staged when publishing the resulting FlowCraft pipelines to a repository. As they are hidden files, they are often overlooked, breaking the execution of the pipelines when run remotely.

There was no reason to keep these files as dotfiles, so they were moved to the resources folder.

### CODE REVIEW

File: flowcraft/generator/engine.py [hunk 1]
```diff
@@ -1382,10 +1384,14 @@ def render_pipeline(self):
...
-                               ".forkTree.json"), "w") as fh:
-            fh.write(json.dumps(self._fork_tree))
+        # Write tree forking information
+        resources_dir = os.path.join(dirname(self.nf_file), "resources")
+
+        if not os.path.exists(resources_dir):
+            os.mkdir(resources_dir)
+        outfile_tree_fork = open(os.path.join(resources_dir, "forkTree.json"), "w")
+        outfile_tree_fork.write(json.dumps(dict_viz))
+        outfile_tree_fork.close()
```
Comment (tiagofilipe12) [inline flowcraft/generator/engine.py]: Hmm interesting that you are keeping consistency here between both methods. But now you can see that some duplication exists between the two methods? Maybe you could write a function called something like `write_dag_to_file` and use it in both methods.
Then you basically put everything inside that function and re-use in both places:

```python
def write_dag_to_file(file_name, dict_viz):
    resources_dir = os.path.join(dirname(self.nf_file), "resources")
    if not os.path.exists(resources_dir):
            os.mkdir(resources_dir)
    outfile_tree_fork = open(os.path.join(resources_dir, file_name), "w")
    outfile_tree_fork.write(json.dumps(dict_viz))
    outfile_tree_fork.close()
```

or you can even go with `with open`... Then you just call the function in both places. Something like:

```python
write_dag_to_file('forkTree.json', dict_viz)
```

File: flowcraft/generator/inspect.py [hunk 2]
```diff
@@ -1465,8 +1465,8 @@ def _prepare_static_info(self):
         return pipeline_files
 
     def _dag_file_to_dict(self):
-        """Function that opens the dotfile named .treeDag.json in the current
-        working directory
+        """Function that opens the accessory named treeDag.json in the
```
Comment (tiagofilipe12) [inline flowcraft/generator/inspect.py]: What is accessory? Also notice that this method not only opens that file but loads its content to a dict and hence the docstring is incomplete. It was already incomplete before I know.

File: flowcraft/generator/templates/downsample_fastq.nf [hunk 3]
```diff
@@ -17,7 +17,7 @@ process downsample_fastq_{{ pid }} {
     {% include "post.txt" ignore missing %}
 
     tag { "${sample_id}" }
-    publishDir "results/downsample_fastq_{{ pid }}/", pattern: "_ss.*"
+    publishDir "results/downsample_fastq_{{ pid }}/", pattern: "*_ss.*"
```
Comment (tiagofilipe12) [inline flowcraft/generator/templates/downsample_fastq.nf]: Wait is this related with this PR? Or is it another fix?

File: flowcraft/generator/templates/downsample_fastq.nf [hunk 4]
```diff
@@ -37,4 +37,4 @@ process downsample_fastq_{{ pid }} {
 
 }
 
-{{ forks }}
\ No newline at end of file
+{{ forks }}
```
Comment (tiagofilipe12) [inline flowcraft/generator/templates/downsample_fastq.nf]: same here

File: flowcraft/templates/downsample_fastq.py [hunk 5]
```diff
@@ -156,8 +157,8 @@ def main(sample_id, fastq_pair, genome_size, depth, clear, seed):
                     os.remove(rp)
 
     else:
-        os.symlink(p1, "{}._ss.fq.gz".format(bn1))
-        os.symlink(p2, "{}._ss.fq.gz".format(bn2))
+        shutil.copy(p1, "{}._ss.fq.gz".format(bn1))
+        shutil.copy(p2, "{}._ss.fq.gz".format(bn2))
```
Comment (tiagofilipe12) [inline flowcraft/templates/downsample_fastq.py]: and here? I would recommend to do this in a different PR. Also in changelog this isn't reflected.

File: flowcraft/generator/templates/downsample_fastq.nf (hunk 3 above)
Comment (cimendes) [reply flowcraft/generator/templates/downsample_fastq.nf]: Ah it was a bug that I found that I commited to the wrong branch... Oh well :stuck_out_tongue:

File: flowcraft/generator/templates/downsample_fastq.nf (hunk 4 above)
Comment (cimendes) [reply flowcraft/generator/templates/downsample_fastq.nf]: I don't get this comment? Same what?

File: flowcraft/tests/test_assemblerflow.py [hunk 6]
```diff
@@ -51,12 +51,13 @@ def test_build_file_2(tmp):
...
 
-    assert sorted(os.listdir(tmp)) == [".forkTree.json", ".treeDag.json",
-                                       "containers.config",
+    assert sorted(os.listdir(tmp)) == ["containers.config",
                                        "lib", "nextflow.config", "params.config",
-                                       "resources.config", "teste.html",
+                                       "resources", "resources.config", "teste.html",
                                        "teste.nf", "user.config"]
 
+    assert sorted(os.listdir(os.path.join(tmp, "resources"))) == ["forkTree.json", "treeDag.json"]
```
Comment (ODiogoSilva) [inline flowcraft/tests/test_assemblerflow.py]: Nice 👍

File: flowcraft/generator/templates/downsample_fastq.nf (hunk 3 above)
Comment (cimendes) [reply flowcraft/generator/templates/downsample_fastq.nf]: Added to changelog

File: flowcraft/templates/downsample_fastq.py (hunk 5 above)
Comment (cimendes) [reply flowcraft/templates/downsample_fastq.py]: added to changelog

File: flowcraft/generator/engine.py [hunk 7]
```diff
@@ -1382,10 +1384,14 @@ def render_pipeline(self):
...
         self.dag_to_file(dict_viz)
 
-        # Write tree forking information for dotfile
-        with open(os.path.join(dirname(self.nf_file),
-                               ".forkTree.json"), "w") as fh:
-            fh.write(json.dumps(self._fork_tree))
+        # Write tree forking information
+        resources_dir = os.path.join(dirname(self.nf_file), "resources")
+
+        if not os.path.exists(resources_dir):
```
[the remark this answers is not in the document]
Comment (cimendes) [reply flowcraft/generator/engine.py]: This function is now general to write both the treedata.json and forktrees.json files. The check was moved to the render_pipeline function

File: flowcraft/flowcraft.py [hunk 8]
```diff
@@ -358,6 +358,10 @@ def build(args):
 
     logger.info(colored_print("Building your awesome pipeline..."))
 
+    # copy template to cwd, to allow for immediate execution
+    if not args.pipeline_only:
+        copy_project(parsed_output_nf)
```
[the remark this answers is not in the document]
Comment (cimendes) [reply flowcraft/flowcraft.py]: comment added

File: flowcraft/generator/engine.py [hunk 9]
```diff
@@ -1382,10 +1384,14 @@ def render_pipeline(self):
...
+            os.mkdir(resources_dir)
+        outfile_tree_fork = open(os.path.join(resources_dir, "forkTree.json"), "w")
```
[the remark this answers is not in the document]
Comment (cimendes) [reply flowcraft/generator/engine.py]: Done

**Comment by cimendes** [comment]
The verification was moved to the render_pipeline function and the function to write the json was made more general to accommodate both forktree.json and treedag.json files. Thanks @ODiogoSilva for pointing out my very silly mistake!

File: flowcraft/generator/engine.py [hunk 10]
```diff
@@ -1300,19 +1300,21 @@ def _set_configurations(self):
...
 
         Parameters
         ----------
         dict_viz: dict
             Tree like dictionary that is used to export tree data of processes
-            to html file and here for the dotfile .treeDag.json
+            to html file and here for the treeDag.json, stored in the resources directory
 
         """
+        resources_dir = os.path.join(dirname(self.nf_file), "resources")
```
[the remark this answers is not in the document]
Comment (cimendes) [reply flowcraft/generator/engine.py]: done

File: flowcraft/generator/engine.py [hunk 11]
```diff
@@ -1379,13 +1381,16 @@ def render_pipeline(self):
 
                 last_of_us[p.lane] = lst[-1]["children"]
 
+        #check if resources dir exists - necessary for dag files
```
Comment (ODiogoSilva) [inline flowcraft/generator/engine.py]:
```suggestion
        # check if resources dir exists - necessary for dag files
```

File: flowcraft/templates/downsample_fastq.py [hunk 12]
```diff
@@ -36,14 +36,15 @@
 
 """
 
-__version__ = "1.0.0"
-__build__ = "30072018"
+__version__ = "1.0.1"
+__build__ = "21062018"
```
Comment (ODiogoSilva) [inline flowcraft/templates/downsample_fastq.py]: time travel?

```suggestion
__build__ = "21062019"
```

Comment (cimendes) [reply flowcraft/templates/downsample_fastq.py]: I'm missing a tardis emoji!
